File: schemakeeper-avro/src/test/scala/schemakeeper/avro/AvroSchemaUtilsTest.scala
package schemakeeper.avro

import org.apache.avro.generic.GenericData
import org.apache.avro.{Schema, SchemaBuilder}
import org.scalatest.{Matchers, WordSpec}

class AvroSchemaUtilsTest extends WordSpec with Matchers {
  "AvroUtils" should {
    "return schema" when {
      "object is string" in {
        val value: String = ""
        val schema = AvroSchemaUtils.getSchema(value)
        assertResult(Schema.create(Schema.Type.STRING))(schema)
      }

      "object is bytes" in {
        val value: Array[Byte] = Array()
        val schema = AvroSchemaUtils.getSchema(value)
        assertResult(Schema.create(Schema.Type.BYTES))(schema)
      }

      "object is int" in {
        val value: Int = 0
        val schema = AvroSchemaUtils.getSchema(value)
        assertResult(Schema.create(Schema.Type.INT))(schema)
      }

      "object is long" in {
        val value: Long = 0
        val schema = AvroSchemaUtils.getSchema(value)
        assertResult(Schema.create(Schema.Type.LONG))(schema)
      }

      "object is float" in {
        val value: Float = 0
        val schema = AvroSchemaUtils.getSchema(value)
        assertResult(Schema.create(Schema.Type.FLOAT))(schema)
      }

      "object is double" in {
        val value: Double = 0
        val schema = AvroSchemaUtils.getSchema(value)
        assertResult(Schema.create(Schema.Type.DOUBLE))(schema)
      }

      "object is boolean" in {
        val value: Boolean = true
        val schema = AvroSchemaUtils.getSchema(value)
        assertResult(Schema.create(Schema.Type.BOOLEAN))(schema)
      }

      "object is null" in {
        val value: AnyRef = null
        val schema = AvroSchemaUtils.getSchema(value)
        assertResult(Schema.create(Schema.Type.NULL))(schema)
      }

      "object is generic" in {
        val schema = SchemaBuilder
          .builder("test_namespace")
          .record("test")
          .fields()
          .requiredString("f1")
          .endRecord()
        val record = new GenericData.Record(schema)
        assertResult(schema)(AvroSchemaUtils.getSchema(record))
      }
    }

    "check if object has primitive schema" when {
      "object is string" in {
        val value: String = ""
        assert(AvroSchemaUtils.isPrimitive(value))
      }

      "object is bytes" in {
        val value: Array[Byte] = Array()
        assert(AvroSchemaUtils.isPrimitive(value))
      }

      "object is int" in {
        val value: Int = 0
        assert(AvroSchemaUtils.isPrimitive(value))
      }

      "object is long" in {
        val value: Long = 0
        assert(AvroSchemaUtils.isPrimitive(value))
      }

      "object is float" in {
        val value: Float = 0
        assert(AvroSchemaUtils.isPrimitive(value))
      }

      "object is double" in {
        val value: Double = 0
        assert(AvroSchemaUtils.isPrimitive(value))
      }

      "object is boolean" in {
        val value: Boolean = true
        assert(AvroSchemaUtils.isPrimitive(value))
      }

      "object is null" in {
        val value: AnyRef = null
        assert(AvroSchemaUtils.isPrimitive(value))
      }

      "object is generic" in {
        val schema = SchemaBuilder
          .builder("test_namespace")
          .record("test")
          .fields()
          .requiredString("f1")
          .endRecord()
        val record = new GenericData.Record(schema)
        assert(!AvroSchemaUtils.isPrimitive(record))
      }
    }

    "throw error while getting schema" in {
      assertThrows[IllegalArgumentException](AvroSchemaUtils.getSchema(Array[String]()))
    }

    "throw error while checking object type" in {
      assertThrows[IllegalArgumentException](AvroSchemaUtils.isPrimitive(Array[String]()))
    }
  }
}

File: schemakeeper-common/src/test/java/schemakeeper/schema/AvroSchemaUtilsTest.java
package schemakeeper.schema;

import org.apache.avro.Schema;
import org.apache.avro.SchemaBuilder;
import org.apache.avro.generic.GenericData;
import org.apache.avro.generic.GenericRecord;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class AvroSchemaUtilsTest {
    @Test
    public void returnStringSchema() {
        String val = "";
        assertEquals(AvroSchemaUtils.getSchema(val), Schema.create(Schema.Type.STRING));
    }

    @Test
    public void returnBytesSchema() {
        byte[] val = new byte[0];
        assertEquals(AvroSchemaUtils.getSchema(val), Schema.create(Schema.Type.BYTES));
    }

    @Test
    public void returnIntSchema() {
        int val = 1;
        assertEquals(AvroSchemaUtils.getSchema(val), Schema.create(Schema.Type.INT));
    }

    @Test
    public void returnLongSchema() {
        long val = 1L;
        assertEquals(AvroSchemaUtils.getSchema(val), Schema.create(Schema.Type.LONG));
    }

    @Test
    public void returnFloatSchema() {
        float val = 1.0f;
        assertEquals(AvroSchemaUtils.getSchema(val), Schema.create(Schema.Type.FLOAT));
    }

    @Test
    public void returnDoubleSchema() {
        double val = 1.0d;
        assertEquals(AvroSchemaUtils.getSchema(val), Schema.create(Schema.Type.DOUBLE));
    }

    @Test
    public void returnBooleanSchema() {
        boolean val = true;
        assertEquals(AvroSchemaUtils.getSchema(val), Schema.create(Schema.Type.BOOLEAN));
    }

    @Test
    public void returnNullSchema() {
        Object val = null;
        assertEquals(AvroSchemaUtils.getSchema(val), Schema.create(Schema.Type.NULL));
    }

    @Test
    public void returnGenericSchema() {
        Schema schema = SchemaBuilder.builder()
                .record("test")
                .fields()
                .optionalString("f1")
                .endRecord();
        GenericRecord val = new GenericData.Record(schema);

        assertEquals(AvroSchemaUtils.getSchema(val), val.getSchema());
    }

    @Test
    public void checkIfStringIsPrimitive() {
        String val = "";
        assertTrue(AvroSchemaUtils.isPrimitive(val));
    }

    @Test
    public void checkIfBytesIsPrimitive() {
        byte[] val = new byte[0];
        assertTrue(AvroSchemaUtils.isPrimitive(val));
    }

    @Test
    public void checkIfIntIsPrimitive() {
        int val = 1;
        assertTrue(AvroSchemaUtils.isPrimitive(val));
    }

    @Test
    public void checkIfLongIsPrimitive() {
        long val = 1L;
        assertTrue(AvroSchemaUtils.isPrimitive(val));
    }

    @Test
    public void checkIfFloatIsPrimitive() {
        float val = 1.0f;
        assertTrue(AvroSchemaUtils.isPrimitive(val));
    }

    @Test
    public void checkIfDoubleIsPrimitive() {
        double val = 1.0d;
        assertTrue(AvroSchemaUtils.isPrimitive(val));
    }

    @Test
    public void checkIfBooleanIsPrimitive() {
        boolean val = true;
        assertTrue(AvroSchemaUtils.isPrimitive(val));
    }

    @Test
    public void checkIfNullIsPrimitive() {
        Object val = null;
        assertTrue(AvroSchemaUtils.isPrimitive(val));
    }

    @Test
    public void checkIfGenericIsNotPrimitive() {
        Schema schema = SchemaBuilder.builder()
                .record("test")
                .fields()
                .optionalString("f1")
                .endRecord();
        GenericRecord val = new GenericData.Record(schema);

        assertFalse(AvroSchemaUtils.isPrimitive(val));
    }

    @Test
    public void throwErrorWhileGettingSchema() {
        String[] val = new String[0];
        assertThrows(IllegalArgumentException.class, () -> AvroSchemaUtils.getSchema(val));
    }

    @Test
    public void throwErrorWhileCheckingThatObjectIsPrimitive() {
        String[] val = new String[0];
        assertThrows(IllegalArgumentException.class, () -> AvroSchemaUtils.isPrimitive(val));
    }
}
